File: EmotionEngine/src/main/java/co/q64/emotion/lang/value/standard/ListValue.java
package co.q64.emotion.lang.value.standard;

import co.q64.emotion.lang.value.Value;
import co.q64.emotion.lang.value.Values;
import co.q64.emotion.types.Operation;
import co.q64.emotion.util.Rational;
import lombok.EqualsAndHashCode;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

@EqualsAndHashCode
public class ListValue implements Value {
    private final List<Value> values;

    private ListValue(List<Value> values) {
        this.values = values;
    }

    public static ListValue of(List<?> in) {
        List<Value> values = new ArrayList<>();
        for (Object object : in) {
            if (in instanceof Value) {
                values.add((Value) object);
            } else {
                values.add(Values.create(object.toString()));
            }
        }
        return new ListValue(values);
    }

    public Value operate(Value value, Operation type) {
        return null;
    }

    public int compareTo(Value value) {
        if (value.isList()) {
            return values.equals(value.iterate()) ? 0 : -1;
        }
        return -1;
    }

    public List<Value> iterate() {
        return new ArrayList<>(values);
    }

    public Rational asNumber() {
        return Rational.of(values.size());
    }

    public boolean asBoolean() {
        return !values.isEmpty();
    }

    @Override
    public boolean isList() {
        return true;
    }

    @Override
    public String toString() {
        return "[" + values.stream().map(Value::toString).collect(Collectors.joining(",")) + "]";
    }
}

File: emotion-core/src/commonMain/kotlin/co/q64/emotion/core/lang/value/ListValue.kt
package co.q64.emotion.core.lang.value

import co.q64.emotion.core.math.rational
import co.q64.emotion.core.value.ListType
import co.q64.emotion.core.value.Value

data class ListValue(
    private val value: List<Value>
) : Value {
    override val type get() = ListType
    override val number get() = value.size.rational()
    override val list: List<Value> get() = value

    override fun compareTo(other: Value): Int {
        TODO("not implemented")
    }

    override fun toString() = value.joinToString(",")
}